### PR TITLE
[hail/ptypes] Copy IRs before inferring

### PR DESCRIPTION
**Pull request by tpoterba**
The `assert(_ptype2 == null)` check in InferPType is breaking here on
certain complex pipelines in a way I don't want to debug.

There's no IR sharing within the IR (see utility I added), but there
must be subtrees that are inferred multiple times in different Compile
calls.

This is a safe stop-gap.

### CODE REVIEW

File: hail/src/main/scala/is/hail/expr/ir/RefEquality.scala [hunk 1]
```diff
@@ -54,3 +54,13 @@ class Memo[T] private(val m: mutable.HashMap[RefEquality[BaseIR], T]) {
   def delete(ir: RefEquality[BaseIR]): Unit = m -= ir
 }
 
+
+object HasIRSharing {
```
Comment (akotlar) [inline hail/src/main/scala/is/hail/expr/ir/RefEquality.scala]: could we use this, or add a test?

File: hail/src/main/scala/is/hail/expr/ir/RefEquality.scala [hunk 2]
```diff
@@ -54,3 +54,13 @@ class Memo[T] private(val m: mutable.HashMap[RefEquality[BaseIR], T]) {
...
+  def apply(ir: BaseIR): Boolean = {
+    val m = mutable.HashSet.empty[RefEquality[BaseIR]]
+
+    def recur(x: BaseIR): Boolean = m.contains(RefEquality(x)) || x.children.exists(recur)
```
Comment (akotlar) [inline hail/src/main/scala/is/hail/expr/ir/RefEquality.scala]: I think if doesn't contain needs

```scala
m.add(RefEquality(x))
```

**Comment by tpoterba** [comment]
yeah, saw the same thing after adding a test, thanks for catching!

File: hail/src/main/scala/is/hail/expr/ir/RefEquality.scala [hunk 3]
```diff
@@ -54,3 +54,22 @@ class Memo[T] private(val m: mutable.HashMap[RefEquality[BaseIR], T]) {
   def delete(ir: RefEquality[BaseIR]): Unit = m -= ir
 }
 
+
+object HasIRSharing {
+  def apply(ir: BaseIR): Boolean = {
+    val m = mutable.HashSet.empty[RefEquality[BaseIR]]
+
+    def recur(x: BaseIR): Boolean = {
+      println(s"$x: $m")
```
Comment (akotlar) [inline hail/src/main/scala/is/hail/expr/ir/RefEquality.scala]: debug :)

File: hail/src/main/scala/is/hail/expr/ir/RefEquality.scala [hunk 4]
```diff
@@ -54,3 +54,22 @@ class Memo[T] private(val m: mutable.HashMap[RefEquality[BaseIR], T]) {
...
+      val re = RefEquality(x)
+      if (m.contains(re))
+        true
+      else {
+        m.add(RefEquality(x))
```
Comment (akotlar) [inline hail/src/main/scala/is/hail/expr/ir/RefEquality.scala]: can be m.add(re)

**Pull request merged**